library/data/buffer/parser.lean

```diff
@@ -188,6 +199,22 @@ def fix_core (F : parser α → parser α) : ∀ (max_depth : ℕ), parser α
 | 0             := failure
 | (max_depth+1) := F (fix_core max_depth)
 
+/-- Matches a digit (0-9). -/
+def digit : parser nat := do
+  c ← any_char,
+  let c' := c.to_nat - '0'.to_nat,
+  if 0 ≤ c' ∧ c' ≤ 9
+    then pure c'
+    else parser.fail $ "expected a digit, got: " ++ c.to_string
+
+/-- Matches a natural number. Large numbers may cause performance issues, so
+don't run this parser on untrusted input. -/
+def nat : parser nat := do
```

Comment:
```suggestion
def nat : parser nat := decorate_error "<number>" $ do
```
Better error-message and backtracking.  (The `<|>` only backtracks if the position doesn't advance.)

Reply:
Added the `decorate_error`. I'm afraid I don't understand the `<|>` comment -- was there supposed to be another suggested change?

Reply:
No this was just (an apparently failed attempt at) an explanation.

Reply:
Ah okay, thanks!